haystack/core/component/sockets.py

```diff
@@ -1,57 +1,106 @@
+    the `Component`.
+
+    Usage:
+    ```python
+    from haystack.components.builders.prompt_builder import PromptBuilder
+
+    prompt_template = \"""
+    Given these documents, answer the question.\nDocuments:
+    {% for doc in documents %}
+        {{ doc.content }}
+    {% endfor %}
+
+    \nQuestion: {{question}}
+    \nAnswer:
+    \"""
 
-class _empty:
-    """Custom object for marking InputSocket.default_value as not set."""
+    prompt_builder = PromptBuilder(template=prompt_template)
+    sockets = {"question": InputSocket("question", Any), "documents": InputSocket("documents", Any)}
+    inputs = Sockets(component=prompt_builder, sockets=sockets, sockets_type=InputSocket)
+    inputs
+    >>> PromptBuilder inputs:
+    >>>   - question: Any
+    >>>   - documents: Any
 
+    inputs.question
+    >>> InputSocket(name='question', type=typing.Any, default_value=<class 'haystack.core.component.types._empty'>, is_variadic=False, senders=[])
+    ```
+    """
 
-@dataclass
-class InputSocket:
-    name: str
-    type: Type
-    default_value: Any = _empty
-    is_variadic: bool = field(init=False)
-    senders: List[str] = field(default_factory=list)
+    # We're using a forward declaration here to avoid a circular import.
+    def __init__(
+        self,
+        component: "Component",  # type: ignore[name-defined] # noqa: F821
+        sockets: Dict[str, SocketsType],
+        sockets_type: SocketsType,
+    ):
+        """
+        Create a new Sockets object.
+        We don't do any enforcement on the types of the sockets here, the `sockets_type` is only used for
+        the `__repr__` method.
+        We could do without it and use the type of a random value in the `sockets` dict, but that wouldn't
+        work for components that have no sockets at all. Either input or output.
+        """
+        self._sockets_type = sockets_type
+        self._component = component
+        self._sockets = sockets
+        self.__dict__.update(sockets)
 
-    @property
-    def is_mandatory(self):
-        return self.default_value == _empty
+    def __setitem__(self, key: str, socket: SocketsType):
+        """
+        Adds a new socket to this Sockets object.
+        This eases a bit updating the list of sockets after Sockets has been created.
+        That should happen only in the `component` decorator.
+        """
+        self._sockets[key] = socket
+        self.__dict__[key] = socket
 
-    def __post_init__(self):
+    def _component_name(self) -> str:
+        if pipeline := getattr(self._component, "__haystack_added_to_pipeline__"):
+            # This Component has been added in a Pipeline, let's get the name from there.
+            return pipeline.get_component_name(self._component)
+
+        # This Component has not been added to a Pipeline yet, so we can't know its name.
+        # Let's use the class name instead.
+        return self._component.__class__.__name__
```

Comment:
I would add something "dynamic" to distinguish between different instances of the same component class, like `<PromptBuilder component at 0x10739e4d0>`. Maybe we can just return `str(self._component)`?